fuel-core/src/schema/balance.rs

```diff
@@ -0,0 +1,192 @@
+use crate::database::{Database, KvStoreError};
+use crate::model::coin::{Coin as CoinModel, CoinStatus};
+use crate::schema::scalars::{Address, AssetId, U64};
+use crate::state::IterDirection;
+use async_graphql::InputObject;
+use async_graphql::{
+    connection::{query, Connection, Edge, EmptyFields},
+    Context, Object,
+};
+use fuel_storage::Storage;
+use itertools::Itertools;
+
+pub struct Balance {
+    owner: fuel_types::Address,
+    amount: u64,
+    asset_id: fuel_types::AssetId,
+}
+
+#[Object]
+impl Balance {
+    async fn owner(&self) -> Address {
+        self.owner.into()
+    }
+
+    async fn amount(&self) -> U64 {
+        self.amount.into()
+    }
+
+    async fn asset_id(&self) -> AssetId {
+        self.asset_id.into()
+    }
+}
+
+#[derive(InputObject)]
+struct BalanceFilterInput {
+    /// Filter coins based on the `owner` field
+    owner: Address,
+}
+
+#[derive(Default)]
+pub struct BalanceQuery;
+
+#[Object]
+impl BalanceQuery {
+    async fn balance(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(desc = "address of the owner")] owner: Address,
+        #[graphql(desc = "asset_id of the coin")] asset_id: AssetId,
+    ) -> async_graphql::Result<Balance> {
+        let db = ctx.data_unchecked::<Database>();
+
+        let coin_ids: Vec<fuel_tx::UtxoId> =
+            db.owned_coins(owner.into(), None, None).try_collect()?;
+        let mut coins: Vec<(fuel_tx::UtxoId, CoinModel)> = coin_ids
```

Comment:
If we combine all these iterator ops into one stream, we'll cut down memory overhead by not requiring all coins to be loaded into memory into a single vec (try_collect), and they could be summed and discarded as they load one by one from the db.